### PR TITLE
Job Summary - Part 1

### PR DESCRIPTION
**Pull request by diptanu**
This PR introduces the job summary table in the state store and updates the summary when jobs and allocations are inserted/updated/deleted.

Subsequent PRs would introduce the API and CLI support for the same.

This PR doesn't account for number of pending allocations, will come in a followup PR.

### CODE REVIEW

File: nomad/state/state_store.go [hunk 1]
```diff
@@ -356,6 +361,17 @@ func (s *StateStore) DeleteJob(index uint64, jobID string) error {
 		return fmt.Errorf("index update failed: %v", err)
 	}
 
+	// Delete the job summary
+	summary, err := s.JobSummary(jobID)
```
Comment (dadgar) [inline nomad/state/state_store.go]: Why do you have to look up the summary first, just delete the job

File: nomad/state/state_store.go [hunk 2]
```diff
@@ -436,6 +452,21 @@ func (s *StateStore) JobsByGC(gc bool) (memdb.ResultIterator, error) {
 	return iter, nil
 }
 
+// JobSummary returns an iterator over al the jobs which matches a specific id.
```
Comment (dadgar) [inline nomad/state/state_store.go]: The comment has a spelling mistake and doesn't particularly make sense

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -436,6 +452,21 @@ func (s *StateStore) JobsByGC(gc bool) (memdb.ResultIterator, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	if existing != nil {
+		return existing.(*structs.JobSummary), nil
+	}
+
+	return nil, nil
+}
+
```
Comment (dadgar) [inline nomad/state/state_store.go]: I would add a Prefix version as well

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -820,6 +856,11 @@ func (s *StateStore) UpsertAllocs(index uint64, allocs []*structs.Allocation) er
 			return fmt.Errorf("alloc insert failed: %v", err)
 		}
 
+		existingAlloc, _ := existing.(*structs.Allocation)
```
Comment (dadgar) [inline nomad/state/state_store.go]: Move the `exist` variable out of the else and reuse that

File: nomad/state/state_store.go [hunk 5]
```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+				Name:     tg.Name,
+				Queued:   tg.Count,
+				Complete: 0,
+				Failed:   0,
+				Running:  0,
+				Starting: 0,
+			}
+			existing.Summary[tg.Name] = newSummary
+		} else {
+			if summary.Queued > tg.Count {
```
Comment (dadgar) [inline nomad/state/state_store.go]: What happens if the count goes up

File: nomad/state/state_store.go [hunk 6]
```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
 	return structs.JobStatusPending, nil
 }
 
+func (s *StateStore) updateSummaryWithJob(job *structs.Job, txn *memdb.Txn) error {
```
Comment (dadgar) [inline nomad/state/state_store.go]: Comment the method

File: nomad/state/state_store.go [hunk 7]
```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+		}
+	}
+
+	if err := txn.Insert("jobsummary", existing); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *StateStore) updateSummaryWithAlloc(newAlloc *structs.Allocation,
```
Comment (dadgar) [inline nomad/state/state_store.go]: Comment the method

File: nomad/state/state_store.go [hunk 8]
```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+	existingAlloc *structs.Allocation, txn *memdb.Txn) error {
+
+	existing, err := s.JobSummary(newAlloc.JobID)
+	if err != nil {
+		return fmt.Errorf("lookup of job summary failed: %v", err)
+	}
+
+	// If we can't find an existing job summary entry then we are not going to create a
+	// new job summary entry for an allocation with that job id since we don't
+	// know the task group counts for that jobA
```
Comment (dadgar) [inline nomad/state/state_store.go]: `jobA`

File: nomad/structs/structs.go [hunk 9]
```diff
@@ -933,6 +933,20 @@ const (
 	CoreJobPriority = JobMaxPriority * 2
 )
 
+type JobSummary struct {
+	JobID   string
+	Summary map[string]TaskGroupSummary
+}
+
+type TaskGroupSummary struct {
+	Name     string
```
Comment (dadgar) [inline nomad/structs/structs.go]: Remove the name, it is stored in a map indexed by name

File: nomad/state/state_store.go [hunk 10]
```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+	if existing == nil {
+		return nil
+	}
+
+	tgSummary, ok := existing.Summary[newAlloc.TaskGroup]
+	if !ok {
+		return nil
+	}
+	if existingAlloc == nil {
+		switch newAlloc.DesiredStatus {
```
Comment (dadgar) [inline nomad/state/state_store.go]: If it is a new allocation, wouldn't the only accepted values be DesiredStatus: Run, Failed and ClientStatus: Pending.

File: nomad/state/state_store.go [hunk 11]
```diff
@@ -1136,6 +1177,109 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+		}
+	}
+
+	existing.Summary[newAlloc.TaskGroup] = tgSummary
+	if err := txn.Insert("jobsummary", existing); err != nil {
+		return fmt.Errorf("inserting job summary failed: %v", err)
+	}
+	return nil
+}
+
```
Comment (dadgar) [inline nomad/state/state_store.go]: Add restore method in StateStore and FSM.

Can see this PR: https://github.com/hashicorp/nomad/pull/540

File: nomad/state/state_store.go (hunk 10 above)
Comment (diptanu) [inline nomad/state/state_store.go]: @dadgar Wasn't sure, wanted to make sure that the states are properly accounted for.

File: nomad/state/state_store.go (hunk 5 above)
Comment (diptanu) [inline nomad/state/state_store.go]: I didn't handle that logic until now.

File: nomad/state/schema.go [hunk 12]
```diff
@@ -121,6 +122,24 @@ func jobTableSchema() *memdb.TableSchema {
 	}
 }
 
+// jobSummarySchema returns the memdb schema for the job summary table
+func jobSummarySchema() *memdb.TableSchema {
+	return &memdb.TableSchema{
+		Name: "jobsummary",
```
Comment (dadgar) [inline nomad/state/schema.go]: `job_summary`

File: nomad/state/state_store.go [hunk 13]
```diff
@@ -455,6 +465,45 @@ func (s *StateStore) JobsByGC(gc bool) (memdb.ResultIterator, error) {
...
+		return nil, err
+	}
+	if existing != nil {
+		return existing.(*structs.JobSummary), nil
+	}
+
+	return nil, nil
+}
+
+// JobSummaries walks the entire job summary table and retuns all the job
```
Comment (dadgar) [inline nomad/state/state_store.go]: retuns

File: nomad/state/state_store.go [hunk 14]
```diff
@@ -727,7 +776,8 @@ func (s *StateStore) Evals() (memdb.ResultIterator, error) {
 	return iter, nil
 }
 
-// UpdateAllocFromClient is used to update an allocation based on input
+// UStarting pdateAllocFromClient is used to update an allocation based on input
```
Comment (dadgar) [inline nomad/state/state_store.go]: ?

File: nomad/state/state_store.go [hunk 15]
```diff
@@ -1155,6 +1213,118 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+	if existing == nil {
+		return nil
+	}
+
+	tgSummary, ok := existing.Summary[newAlloc.TaskGroup]
+	if !ok {
+		return nil
+	}
+	if existingAlloc == nil {
+		switch newAlloc.DesiredStatus {
```
Comment (dadgar) [inline nomad/state/state_store.go]: Only use client status

File: nomad/state/state_store.go [hunk 16]
```diff
@@ -1155,6 +1213,118 @@ func (s *StateStore) getJobStatus(txn *memdb.Txn, job *structs.Job, evalDelete b
...
+		case structs.AllocClientStatusPending:
+			tgSummary.Starting += 1
+		case structs.AllocClientStatusRunning:
+			tgSummary.Running += 1
+		}
+		if tgSummary.Queued > 0 {
+			tgSummary.Queued -= 1
+		}
+	} else if existingAlloc.ClientStatus != newAlloc.ClientStatus {
+		// Incrementing the clint of the bin of the current state
```
Comment (dadgar) [inline nomad/state/state_store.go]: Spelling

File: nomad/structs/structs.go [hunk 17]
```diff
@@ -345,6 +345,10 @@ type PeriodicForceRequest struct {
 	WriteRequest
 }
 
+type ServerMembersRequest struct {
```
Comment (dadgar) [inline nomad/structs/structs.go]: Comment

File: nomad/structs/structs.go [hunk 18]
```diff
@@ -937,6 +941,28 @@ const (
...
+type TaskGroupSummary struct {
+	Queued   int
+	Complete int
+	Failed   int
+	Running  int
+	Starting int
+	Lost     int
+}
+
+// Total returns the total number of allocations for the task group.
```
Comment (dadgar) [inline nomad/structs/structs.go]: Remove

**Pull request merged**